### PR TITLE
[WIP] Improved multilateration script

### LINKED ISSUE #148: PASSIVE SONAR: improve multilateration algorithm
The multilateration in mil_passive_sonar is fairly naive. Explore methods other than least squares or a more guarded version of least squares.

### PR DESCRIPTION
**Pull request by jaxnb**
closes #148

Improved multilateration script

Needs testing on boat

### CODE REVIEW

File: drivers/mil_passive_sonar/scripts/multilateration [hunk 1]
```diff
@@ -57,50 +66,61 @@ class MultilaterationNode(object):
         except tf2_ros.TransformException, e:
             rospy.logwarn('TF Exception: {}'.format(e))
             return
+        if self.last_pos is not None and np.linalg.norm(self.last_pos - origin) < 0.2:
```
Comment (kev-the-dev) [inline drivers/mil_passive_sonar/scripts/multilateration:69]: the `0.2` magic number should be a ROS param

File: drivers/mil_passive_sonar/scripts/multilateration [hunk 2]
```diff
@@ -24,30 +24,39 @@ class MultilaterationNode(object):
...
         self.enable_srv = rospy.Service('~enable', SetBool,
                 self.enable_cb)
+        self.reset_srv = rospy.Service('~reset', Trigger,
+                                       self.reset_cb)
 
     def enable_cb(self, req):
-        if req.data is not self.enabled:
-            self.vec_samples.clear()
+        #if req.data is not self.enabled:
+        #    self.vec_samples.clear()
```
Comment (kev-the-dev) [inline drivers/mil_passive_sonar/scripts/multilateration:45]: Can remove dead code

File: drivers/mil_passive_sonar/scripts/multilateration [hunk 3]
```diff
@@ -24,30 +24,39 @@ class MultilaterationNode(object):
...
+        self.last_pos = None
+        self.global_frame = 'enu'
         self.heading_sub = rospy.Subscriber('/hydrophones/direction',
                 Vector3Stamped, self.heading_cb, queue_size=10)
         self.position_pub = rospy.Publisher('/hydrophones/position',
                 PointStamped, queue_size=1)
         self.enable_srv = rospy.Service('~enable', SetBool,
                 self.enable_cb)
+        self.reset_srv = rospy.Service('~reset', Trigger,
+                                       self.reset_cb)
```
Comment (kev-the-dev) [inline drivers/mil_passive_sonar/scripts/multilateration:41]: The txros interface should be changed to provide a function to call reset. I believe this was added at lakeday but not committed

File: drivers/mil_passive_sonar/scripts/multilateration [hunk 4]
```diff
@@ -57,50 +66,61 @@ class MultilaterationNode(object):
...
-        for x, y in zip(begin_pts, perp_unit_vecs):
-            begin = x.reshape(2, 1)
-            perp_vec = y.reshape(2, 1)
-            A = perp_vec.dot(perp_vec.T)
-            Ap = A.dot(begin)
-            A_sum += A
-            Ap_sum += Ap
-
-        return np.linalg.inv(A_sum).dot(Ap_sum)
+    def ls_line_intersection2d(self, lines_array):
```
Comment (DSsoto) [inline drivers/mil_passive_sonar/scripts/multilateration:87]: Just want to point out that what you're doing here (average of intersections of all possible pairs) yields a mathematically suboptimal estimate of the position of the source. Throwing out observations that are too close is sensible but swapping out least squares intersection for this average is not. Consider using this as a preprocessing step to filter the array of lines.

**Comment by jaxnb** [comment]
Messed up my git stuff, see #164

**Pull request closed without merging**